kotlin/internal/jvm/compile.bzl

```diff
@@ -641,7 +880,22 @@ def _kt_jvm_produce_output_jar_actions(
     annotation_processors = _plugin_mappers.targets_to_annotation_processors(ctx.attr.plugins + ctx.attr.deps)
     ksp_annotation_processors = _plugin_mappers.targets_to_ksp_annotation_processors(ctx.attr.plugins + ctx.attr.deps)
     transitive_runtime_jars = _plugin_mappers.targets_to_transitive_runtime_jars(ctx.attr.plugins + ctx.attr.deps)
-    plugins = _new_plugins_from(ctx.attr.plugins + _exported_plugins(deps = ctx.attr.deps))
+
+    # Automatically add KAPT plugin when there are annotation processors
+    all_plugins = ctx.attr.plugins + _exported_plugins(deps = ctx.attr.deps)
+
+    # Check if KAPT plugin is already present by looking for its ID
```

Comment:
The general pattern is to build the kapt dependency into the toolchain/rule dependencies, rather than adding here.